lib/multi_channel.ml

```diff
@@ -76,13 +78,16 @@ let init_domain_state mchan dls_state =
   [@@inline never]
 
 let get_local_state mchan =
-  let dls_state = Domain.DLS.get dls_key in
-  if dls_state.id >= 0 then dls_state
+  let dls_state = Domain.DLS.get mchan.dls_key in
+  if dls_state.id >= 0 then begin
+    assert (dls_state.id < Array.length mchan.channels);
```

Comment:
The assert is going to execute every time which is sad, but I guess it can stay if you are strongly in favour.